sympy/vector/functions.py

```diff
@@ -6,6 +6,101 @@
 from sympy import diff, integrate, S, simplify
 from sympy.core import sympify
 from sympy.vector.dyadic import Dyadic
+from math import isnan
+
+
+def extended_express(expr,coordsys):
```

Comment:
It would be better to extend the original `express` function. User should not have to choose between two functions.

Reply:
I agree that this would be better long term. However the current express has a focus towards dyadics and inertial rotations. Writing a composite express that manages the cross overs and edge cases for combining these cases is likely to become very messy very quickly. I am not in a position to do all of that right now. 

Implementing it separately ensures the user actually wants to use the conversion to/from cartesian/non-cartesian and it doesn't accidentally corrupt the existing installed codebase. (your comment in february on #17507 that express appears to have been designed for rectilinear coordinates only suggests this is the case, changing express would likely impact on these cases)

I could attempt a 'detect the frame of the vector, and the new frame and if there is a conversion between inertial/non-inertial, use this function instead of the normal one' patch to express if you like, but this is a rather horrible solution and hard to maintain. It would also produce hard to debug weird behaviour for user code that intermittently triggered over it.

Since this functionality is needed (eg  #13175, still open since 2017) , it seems to me that a workable solution for the common use cases is the way to go. Non-cartesian systems can get quickly mathematically complex as they are intrinsically non-linear. Development could easily get mired in discussion of edge cases (the functionality seems to be already partially implemented in the code, but development seems to have halted). 

This code is not a perfect solution, however it can easily be built on in the future and supplies expected and needed sympy functionality now.

Reply:
I would try to add code to `express` that would detect the presence of non-cartesian systems and then branch to `extended_express`. That could then be left as an independent function, but I would not make it public by adding to `__init__`. `express` could be the common external interface though it would still be possible to import `extended_express` explicitly. That way, it would be easier to update the code without breaking user code.

Reply:
The difficulty is the reliable detection of non-cartesian systems with the current CoordSys3D structure, it doesn't afaics have any clear flag to show non-cartesians, and the lambda routines are used by rotations between cartesian frames (see commit  856336bf93.). A flag could be added to CoordSys3D to enable this to be tested for. I am nervous about modifying core functionality for the whole code base. Would you like me to implement that, or is there a better way? Also what do you think of the express patch as stands?

Reply:
> A flag could be potentially be added to CoordSys3D to enable this to be tested for.

That would probably be a viable solution.